### ISSUE TITLE
Timeouts are hidden as "non-compliant scrape target sending blank Content-Type" error

### ISSUE THREAD

**Issue by roidelapluie** [prometheus/prometheus#15337]
### What did you do?

run prometheus

### What did you expect to see?

ts=2024-11-05T10:57:11.005Z caller=scrape.go:1406 level=debug component="scrape manager" scrape_pool=node2 target=http://10.0.0.1:9090/metrics msg="Scrape failed" err="Get \"http://10.0.0.1:9090/metrics\": context deadline exceeded"

### What did you see instead? Under which circumstances?

time=2024-11-05T11:53:49.900+01:00 level=ERROR source=scrape.go:1558 msg="Failed to determine correct type of scrape target." component="scrape manager" scrape_pool=node5 target=http://10.0.0.1:9090/metrics content_type="" fallback_media_type="" err="non-compliant scrape target sending blank Content-Type and no fallback_scrape_protocol specified for target"

### Prometheus version

```text
prometheus, version 3.0.0-beta.1 (branch: main, revision: 33a3aebfc4b9e98db3aca45e0b62e76f454dfde4)
  build user:       roidelapluie@nixos
  build date:       19800101-00:00:00
  go version:       go1.23.0
  platform:         linux/amd64
  tags:             netgo,builtinassets,stringlabels
```

**Comment by egoroof** [comment]
upgrade from beta1 to rc0 brings this. a lot of spam in console. it's because of one target is not available (down), but it interprets as empty content-type header

```
time=2024-11-06T04:17:24.811+07:00 level=ERROR source=/app/scrape/scrape.go:1558 msg="Failed to determine correct type of scrape target." component="scrape manager" scrape_pool=web_servers target=http://localhost:8091/metrics content_type="" fallback_media_type="" err="non-compliant scrape target sending blank Content-Type and no fallback_scrape_protocol specified for target"
time=2024-11-06T04:17:24.812+07:00 level=WARN source=/app/scrape/scrape.go:1447 msg="Append failed" component="scrape manager" scrape_pool=web_servers target=http://localhost:8091/metrics err="non-compliant scrape target sending blank Content-Type and no fallback_scrape_protocol specified for target"
...
100500+ same messages
...
```

**Comment by machine424** [comment]
Also, I didn't follow https://github.com/prometheus/prometheus/pull/15136 closely, but I think Prometheus' `/metrics` should lead the way and send the appropriate content header so users don't have to set `fallback_scrape_protocol` to its scrape definition.

**Comment by alexgreenbank** [comment]
> Why is this tagged as 3.0 ? Do you believe it is a regression from 2.x ?

#15136 was a last minute breaking change added to 3.0 only.

**Comment by alexgreenbank** [comment]
> ~Also, I didn't follow #15136 closely, but I think Prometheus' `/metrics` should lead the way and send the appropriate content header so users don't have to set `fallback_scrape_protocol` to its scrape config.~ (Weird, I had such errors even when the target was reachable and I managed to get rid of them by setting `fallback_scrape_protocol`)

Do you have some details of the targets you were scraping to require the override?

**Comment by machine424** [comment]
> > ~Also, I didn't follow #15136 closely, but I think Prometheus' `/metrics` should lead the way and send the appropriate content header so users don't have to set `fallback_scrape_protocol` to its scrape config.~ (Weird, I had such errors even when the target was reachable and I managed to get rid of them by setting `fallback_scrape_protocol`)
> 
> Do you have some details of the targets you were scraping to require the override?

Can't really say, maybe I was doing sth wrong, maybe that log was just hiding the real issue as this issue describes, I'll let you know if I get that again.

**Comment by simonpasquier** [comment]
One can easily reproduce with the following configuration

```
- job_name: prometheus
  scrape_interval: 5s
  static_configs:
   - targets: ['doesnotexist.example.com']
````

The Prometheus logs show

````
time=2024-11-06T16:10:19.084+01:00 level=ERROR source=/home/simon/workspace/src/github.com/prometheus/prometheus/scrape/scrape.go:1558 msg="Failed to determine correct type of scrape target.
" component="scrape manager" scrape_pool=prometheus target=http://doesnotexist.example.com/metrics content_type="" fallback_media_type="" err="non-compliant scrape target sending blank Conte
nt-Type and no fallback_scrape_protocol specified for target"
time=2024-11-06T16:10:19.084+01:00 level=ERROR source=/home/simon/workspace/src/github.com/prometheus/prometheus/scrape/scrape.go:1558 msg="Failed to determine correct type of scrape target.
" component="scrape manager" scrape_pool=prometheus target=http://doesnotexist.example.com/metrics content_type="" fallback_media_type="" err="non-compliant scrape target sending blank Conte
nt-Type and no fallback_scrape_protocol specified for target"
````

I agree with @roidelapluie that the issue is at https://github.com/prometheus/prometheus/blob/37f3f3f2db95539a9c84242936018908ffb3dd50/scrape/scrape.go#L1437

In the case where the scraper couldn't connect to the target the contentType is always empty. And there are several places in scrape/scrape.go where we call `append()` with an empty contentType to insert stale markers.

https://github.com/prometheus/prometheus/blob/37f3f3f2db95539a9c84242936018908ffb3dd50/scrape/scrape.go#L1444
https://github.com/prometheus/prometheus/blob/37f3f3f2db95539a9c84242936018908ffb3dd50/scrape/scrape.go#L1523

**Comment by alexgreenbank** [comment]
I have a fix, just working on the tests. The fix stops logging any errors if the payload to `sl.append()` is empty, so it will continue to allow stale markers to be inserted without filling up the logs.

**Comment by ndTEC** [comment]
having the same problem when scraping ActiveMQ Artemis metrics

**Comment by ii343hbka** [comment]
Same error on scraping yandex/odyssey
But I'm sure that this is not a timeout issue, because curl /metrics costs just around 50ms

**Comment by egoroof** [comment]
the actual issue about timeouts fixed in [3.0.0-rc.1](https://github.com/prometheus/prometheus/releases/tag/v3.0.0-rc.1)

**Comment by insider89** [comment]
Have same issue on 3.0.0. Opened a new issue #15485.
